sui_types/src/gas.rs

```diff
@@ -19,75 +25,304 @@ macro_rules! ok_or_gas_error {
-        format!("Gas balance is {balance}, not enough to pay {amount}")
-    )?;
-    let new_gas_coin = GasCoin::new(*gas_coin.id(), gas_object.version(), balance - amount);
-    let move_object = gas_object.data.try_as_move_mut().unwrap();
-    move_object.update_contents(bcs::to_bytes(&new_gas_coin).unwrap());
-    Ok(())
+/// ComputationCost is a newtype wrapper of InternalGasUnits
+/// to ensure a value of this type is used specifically for computation cost.
+/// Anything that does not change the amount of bytes stored in the authority data store
+/// will charge ComputationCost.
+struct ComputationCost(InternalGasUnits<GasCarrier>);
+
+impl ComputationCost {
+    /// Some computations are also linear to the size of data it operates on.
+    pub fn with_size(&self, size: usize) -> Self {
+        Self(self.0.mul(AbstractMemorySize::new(size as u64)))
+    }
+}
+
+/// StorageCost is a newtype wrapper of InternalGasUnits
+/// to ensure a value of this type is used specifically for storage cost.
+/// Anything that changes the amount of bytes stored in the authority data store
+/// will charge StorageCost.
+struct StorageCost(InternalGasUnits<GasCarrier>);
+
+impl StorageCost {
+    pub fn with_size(&self, size: usize) -> Self {
+        Self(self.0.mul(AbstractMemorySize::new(size as u64)))
+    }
+}
+
+/// A list of constant costs of various operations in Sui.
+struct SuiCostTable {
+    /// A flat fee charged for every transaction. This is also the mimmum amount of
+    /// gas charged for a transaction.
+    pub min_transaction_cost: ComputationCost,
+    /// Computation cost per byte charged for package publish. This cost is primarily
+    /// determined by the cost to verify and link a package. Note that this does not
+    /// include the cost of writing the package to the store.
+    pub package_publish_per_byte_cost: ComputationCost,
+    /// Per byte cost to read objects from the store. This is computation cost instead of
+    /// storage cost because it does not change the amount of data stored on the db.
+    pub object_read_per_byte_cost: ComputationCost,
+    /// Per byte cost to write objects to the store. This is computation cost instead of
+    /// storage cost because it does not change the amount of data stored on the db.
+    pub object_mutation_per_byte_cost: ComputationCost,
+    /// Cost to use shared objects in a transaction, which requires full consensus.
+    pub consensus_cost: ComputationCost,
+
+    /// Unit cost of adding a new byte into the storage.
+    pub new_owned_storage_per_byte_cost: StorageCost,
+    /// Unit cost of making objects shared, per byte. At the moment, we don't distinguish
+    /// the cost of introducing an immutable shared object vs. mutable shared object, since
+    /// they are both objects stored forever.
+    /// TODO: If we ever allow for deleting mutable shared objects, then the cost here
+    /// can be different.
+    pub new_shared_storage_per_byte_cost: StorageCost,
```

Comment:
It's not immediately clear to me that this should be different than `new_owned_storage_per_byte_cost`, since someone can also choose never to free an owned object. Is the idea to add an extra disincentive for shared storage (beyond the ability to never get a rebate)?